src/infra/core/src/ingest/fetch_service/core.rs

```diff
@@ -345,3 +317,87 @@ impl std::convert::From<EventTimeSourceError> for PollingIngestError {
 }
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[cfg(test)]
+mod tests {
+    use std::assert_matches::assert_matches;
+
+    use kamu_datasets::{
+        DatasetEnvVarEncryptionError,
+        DatasetEnvVarValue,
+        SAMPLE_DATASET_ENV_VAR_ENCRYPTION_KEY,
+    };
+    use kamu_datasets_services::DatasetKeyValueServiceSysEnv;
+    use random_names::get_random_name;
+
+    use super::*;
+
+    fn var_and_template() -> (String, String) {
+        let var = get_random_name(Some("var_"), 10);
+        let template = ["${{ env.", &var, " }}"].concat();
+        (var, template)
+    }
+
+    fn ds_env_var(key: &str, val: &str) -> Result<DatasetEnvVar, DatasetEnvVarEncryptionError> {
+        DatasetEnvVar::new(
+            key,
+            Utc::now(),
+            &DatasetEnvVarValue::Regular(val.to_string()),
+            &DatasetID::new_seeded_ed25519(b"foo"),
+            SAMPLE_DATASET_ENV_VAR_ENCRYPTION_KEY,
+        )
+    }
+
+    #[test]
+    fn test_fetch_service_template_string() {
+        let temp_dir = tempfile::tempdir().unwrap();
+
+        let fetch_svc = FetchService::new(
+            Arc::new(ContainerRuntime::default()),
+            None,
+            None,
+            None,
+            None,
+            Arc::new(DatasetKeyValueServiceSysEnv::new()),
+            Arc::new(RunInfoDir::new(temp_dir.path().join("run"))),
+        );
+
+        let (var, template) = var_and_template();
+
+        let env_vars_empty: HashMap<String, DatasetEnvVar> = HashMap::new();
+        let env_vars: HashMap<String, DatasetEnvVar> =
+            HashMap::from([(var.clone(), ds_env_var(&var, "222").unwrap())]);
+
+        // When dataset_env_vars is empty and system env var is set,
+        // its value is being used.
+        std::env::set_var(&var, "111");
```

Comment:
It's not a good practice to mess with the global state (process environment) within the tests. I see that unfortunately one of our old tests does it too.

Since this test doesn't really care where the value comes from (from env vars or elsewhere) it might be much better to mock the `DatasetKeyValueService` to return a fixed set of values.

Reply:
Actually it seems that you don't even need to check this logic here and can avoid sys env altogether.